### PR TITLE
feat(cli) Pass token via cli

### LINKED ISSUE #7292: Support passing token and api server to the cli
### Feature Request
It should be possible for us to pass the token and apiserver info to the cli to target execution of commands via tokens and pointing specific clusters or kube masters respectively, as fashioned below
`helm install stable/mariadb --kube-token 1234aasfsadf --kube-server api.openshift.com/1234/qwer:6443`

### PR DESCRIPTION
**Pull request by waveywaves**
Signed-off-by: Vibhav Bobade <vibhav.bobde@gmail.com>

- Add the ability to pass the token via a cli

fixes https://github.com/helm/helm/issues/7292

### CODE REVIEW

**Comment by marckhouzam** [comment]
This is not my area, but I was trying to figure out if this new global flag may need to be handled in the completion logic.

@waveywaves Could you describe in which use case a user would specify this flag? Am I right to assume it is to authenticate with kube cluster?

**Comment by sbose78** [comment]
I would find this useful. Thank you!

How's the kubeconfig context / API Server specification going to work with this? I'm guessing that's picked up from the default kubeconfig context.

In addition, is it possible to support something similar to the following for `helm` commands as well:
```
oc get pods --token=rQ3BfJ7T1dKWGUK7NH-z2Weu4Dcr8PdCqAAFXEFpRZI --server=https://api.gamore.devcluster.openshift.com:6443
```

**Comment by waveywaves** [comment]
@marckhouzam Yes it will be used to authenticate as well as allow `helm` to be used as middleware for executing commands via users tokens, without having to do multiple kubecontext changes (which is not possible in multi-tenant environments).

**Comment by marckhouzam** [comment]
Thanks @waveywaves. So, I'm thinking about two aspects since this is a global flag:
1. completion
1. plugins

For completion since this flag will affect how to connect with the cluster, it needs to be added to the completion override flag list here:
https://github.com/helm/helm/blob/0fd76b2a2bca5b3febf6488c860a7c6502c70086/cmd/helm/root.go#L34

For example, if you type
```
helm --token 12345 status <TAB>
```
the completion script will run a `helm list` to figure out which releases can be listed for `helm status`.  However, to connect to the cluster `helm list` also needs the token.  By adding it to the list mentioned above, all this will be done automatically.

About plugins, should the token be made available for plugins to use?  All the other global flags have a corresponding environment variable which allows to pass the value of the flag to the plugin. See
https://github.com/helm/helm/blob/0fd76b2a2bca5b3febf6488c860a7c6502c70086/pkg/cli/environment.go#L101
https://github.com/helm/helm/blob/0fd76b2a2bca5b3febf6488c860a7c6502c70086/pkg/cli/environment.go#L64
https://github.com/helm/helm/blob/0fd76b2a2bca5b3febf6488c860a7c6502c70086/cmd/helm/load_plugins.go#L130

**Comment by sbose78** [comment]
Tested this out in a pod in an OpenShift cluster. Looks good.

Verification: 
On doing a  `helm install ... --token=xyz --namespace=xyz` , I could successfully create objects which the `xyz` token had privileges to, but the `default` service account didn't.

File: pkg/kube/config.go
Comment (bacongobbler) [inline pkg/kube/config.go]: Users consuming `kube.GetConfig` would have to change their function signatures once they upgrade. Unfortunately, this would be considered a backwards-incompatible change, so we cannot accept it as it currently stands.

One way to work around this would be to revert this change, and just make `RestClientGetter` set the token parameter directly. Would you mind making that change?

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: pkg/kube/config.go
Comment (waveywaves) [reply pkg/kube/config.go]: @bacongobbler thanks for checking this out I will make that change then.

Comment (waveywaves) [reply pkg/kube/config.go]: @bacongobbler I have made the necessary changes to check it out

**Comment by waveywaves** [comment]
@sbose78 Thank you for that, it does make more sense to have a --server option alongside token as well. Just added that to the PR as well.

**Comment by waveywaves** [comment]
@marckhouzam I have added correspoding Env Vars for the `token` and `server` flags, so they can be consumed by the plugins as well. Let me know if I need to make any further changes

File: pkg/cli/environment.go
Comment (marckhouzam) [inline pkg/cli/environment.go]: kubectl has the `-s` shortflag for `--server`.  Do we want to do the same?

Comment (bacongobbler) [reply pkg/cli/environment.go]: This would prevent other subcommands from using `-s` as a shortflag, right? If so that would block other existing shortflags like `helm create -s/--starter`

Comment (marckhouzam) [reply pkg/cli/environment.go]: Ouch! Nice save.  Thanks!

**Comment by marckhouzam** [comment]
Nice patch.
No more comments from me.
I'll let another maintainer approve though.  I'm still too green 😄

**Comment by karuppiah7890** [comment]
@waveywaves Could you link the issue that this PR is for? 😄

File: pkg/cli/environment.go
Comment (bacongobbler) [inline pkg/cli/environment.go]: should be `kube-token` as per the existing kube config feature flag convention.

File: pkg/cli/environment.go
Comment (bacongobbler) [inline pkg/cli/environment.go]: Can we rename this to `kube-server` as per standard convention? e.g. `--kube-context`, `--kubeconfig`

This way we can differentiate between what comes from Helm, and what comes from Kubernetes.

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: pkg/cli/environment.go
Comment (marckhouzam) [reply pkg/cli/environment.go]: Valid point.
In that case we'll need to adapt the flag when passing it to kubectl during completion.  We'll need to adapte the following method:
https://github.com/helm/helm/blob/12e0ab7e44eb02738e1128ac01ee0843e47b4e79/cmd/helm/root.go#L62
@waveywaves This may be tricky, so let me know if you need help.

File: pkg/cli/environment.go
Comment (marckhouzam) [reply pkg/cli/environment.go]: Valid point.
In that case we'll need to adapt the flag when passing it to kubectl during completion.  We'll need to adapt the following method:
https://github.com/helm/helm/blob/12e0ab7e44eb02738e1128ac01ee0843e47b4e79/cmd/helm/root.go#L62
@waveywaves This may be tricky, so let me know if you need help.

BTW, I'm working on making all this automatic, but it is not quite ready yet.

File: pkg/cli/environment.go
Comment (waveywaves) [reply pkg/cli/environment.go]: @bacongobbler @marckhouzam Thanks for reviewing this. I will make the necessary changes.

File: pkg/cli/environment.go
Comment (waveywaves) [reply pkg/cli/environment.go]: @bacongobbler @marckhouzam Thanks for reviewing this. I will make the necessary changes.

**Comment by waveywaves** [comment]
@marckhouzam @bacongobbler I have made the necessary changes.

**Comment by bacongobbler** [comment]
Hi @waveywaves, it’s the holiday season right now. Probably won’t get to reviewing this until the second week of January. Thanks for your patience.

**Comment by waveywaves** [comment]
@bacongobbler No worries. Have a good one :)

**Comment by bacongobbler** [comment]
closes #5159

**Comment by waveywaves** [comment]
@bacongobbler can you check this out ?

File: cmd/helm/root.go
Comment (marckhouzam) [inline cmd/helm/root.go]: I haven't tested it but I believe this won't work. Say the flag is --kube-server, the first echo will add it unchanged to the helm command used for completion, the second echo will add it unchanged a second time, and the third echo will add it a third time but properly renamed.

Instead you should probably do a single echo with a sed command replacing all three flags.  You can use the form
```
sed -e 's/a/b/' -e 's/c/d' ...
```

File: cmd/helm/root.go
Comment (marckhouzam) [inline cmd/helm/root.go]: What is this git add? 🙂

File: pkg/cli/environment.go
Comment (marckhouzam) [inline pkg/cli/environment.go]: I wonder if we want the variable to be HELM_KUBEAPISERVER?

I'll let @bacongobbler decide.

**Comment by marckhouzam** [comment]
@waveywaves you should be able to run `make test-acceptance` to run some regression tests which include completion tests. Or to focus on completion `make test-acceptance-completion`

**Comment by marckhouzam** [comment]
Hi @waveywaves .  Good news.  With the merging of #7323 , you no longer need to make modifications for completion in this PR.  Everything should be handled automatically for completion.  Therefore, you just need to remove your completion changes and rebase on master.

Thanks.

**Comment by marckhouzam** [comment]
Hi @waveywaves .  Good news.  With the merging of #7323 , you no longer need to make modifications for completion in this PR.  Everything should be handled automatically for completion.  Therefore, you just need to remove your completion changes and rebase on master.

Thanks.

**Comment by waveywaves** [comment]
@marckhouzam That's good news. I will do the same.

File: pkg/cli/environment.go
Comment (waveywaves) [reply pkg/cli/environment.go]: I am updating to KUBEAPISERVER definitely makes more sense

**Comment by bacongobbler** [comment]
This is looking great!

I want to make sure that passing an empty string for `config.KubeToken` and `config.KubeApiServer` won't accidentally pass along some extra headers, tricking certain Kubernetes api server implementations.

Marking this for 3.2 for review

**Comment by 01mcgrady01** [comment]
I tried 
 `helm install happy-panda stable/mariadb --kube-token xxx --kube-apiserver xxx` 
but got error
`unknown flag: --kube-apiserver, --kube-token`
Is that right way to specify k8s cluster token and api server address?

**Comment by bacongobbler** [comment]
yes. Are you sure you're using a version of helm built using this PR?

**Comment by 01mcgrady01** [comment]
> yes. Are you sure you're using a version of helm built using this PR?

I build helm from source code of https://github.com/waveywaves/helm/tree/update/token

File: pkg/cli/environment.go
Comment (waveywaves) [inline pkg/cli/environment.go]: @bacongobbler I have updated this to have empty string checks. This should avoid any empty string security flaws the API might have.

Comment (waveywaves) [reply pkg/cli/environment.go]: > This is looking great!
> 
> I want to make sure that passing an empty string (`&""` instead of `nil`) for `clientConfig.BearerToken` and `clientConfig.ApiServer` won't accidentally pass along some extra headers, tricking certain Kubernetes api server implementations.
> 
> Marking this for 3.2 for review

@bacongobbler This should solve the empty string issue.

File: pkg/cli/environment.go
Comment (bacongobbler) [inline pkg/cli/environment.go]: generally speaking, `if s.KubeToken != ""` is the more idiomatic approach. This looks good though, and we can address it in a follow-up PR.

**Review by bacongobbler — APPROVED** [review]

**Comment by bacongobbler** [comment]
thanks!

File: pkg/cli/environment.go
Comment (bacongobbler) [reply pkg/cli/environment.go]: https://github.com/helm/helm/pull/7721

**Comment by kkorada** [comment]
I am unable to use server and token with helm cli
e.g.
`helm install happy-panda stable/mariadb --kube-token xxx --kube-apiserver xxx`

**Comment by chancejin** [comment]
Same issue with kkorada.

I tried to use helm command in my local.

1. helm list  --kubeconfig {kubeconfigFilePath} --kube-context {contextName}
>> It works fine.

2. helm list --kube-apiserver https://{API server IP}:{API server Port} --kube-token {Kubetoken}
>> It shows error like below
Error: Kubernetes cluster unreachable: Get https://10.251.205.11:6443/version?timeout=32s: x509: certificate signed by unknown authority"

Is there necessary data do I have to add to use "--kube-apiserver" without "--kubeconfig"?

**Comment by bacongobbler** [comment]
Please open a new ticket.

**Pull request merged**
